// sentiment/lib/index.js

var extend = require('extend-object');
var assert = require('assert');
var negations = require('../build/negation');


function tokenize (input) {
  return input
            .replace(/[^a-zA-Z- ]+/g, '')
            .replace('/ {2,}/',' ')
            .toLowerCase()
            .split(' ');
}


module.exports = function (phrase, options, callback) {
  // argumentos
  if (typeof phrase === 'undefined') phrase = '';
  if (typeof options === 'undefined') options = null;
  if (typeof options === 'function') callback = options;
  if (typeof callback === 'undefined') callback = null;

  options = options || {};
  var lang = options.lang || 'en';

  var negationList = negations[lang] || [];
  var afinn = {};
//diccionario según lengua
  if (!options.strict) {
    try {
      var filename = ['AFINN', lang, 'json'].join('.');
      afinn = require('../build/' + filename);
    } catch (e) {
      throw new Error('Language \'' + lang + '\' not supported.');
    }
  }
/*
  if (options.category) {
    assert.equal(typeof options.category, 'string',
      "property 'options.category' must be a string");
    try {
      filename = ['AFINN', options.category, lang, 'json'].join('.');
      var categoryData = require('../build/' + filename);
      afinn = extend(afinn, categoryData);
    } catch (e) {
      throw new Error("Category '" + options.category + "' for language '" +
        lang + "' not supported.");
    }
  }*/

  // Merge
  if (options.inject) {
    assert.equal(typeof options.inject, 'object',
      "property 'options.inject' must be an object.");
    afinn = extend(afinn, options.inject);
  }

  // guardamos los tokens
  var tokens      = tokenize(phrase),
      score       = 0,
      words       = [],
      positive    = [],
      negative    = [],
      negation    = [];

  // buscamos negaciones
  var len = tokens.length;
  while (len--) {
    var obj = tokens[len];
    if (negationList.indexOf(obj) !== -1) {
      negation.push(obj);
    }
  }

  // iteremos los tokens
    //tratamos las negaciones
    //buscamos las palabras del mensaje en los 
    //diccionarios y sumamos el valor
  if (!negation.length) {
    len = tokens.length;
    while (len--) {
      var obj = tokens[len];
      if (negationList.indexOf(obj) !== -1) {
        negation.push(obj);
      }
      var item = afinn[obj];
      if (!afinn.hasOwnProperty(obj)) continue;

      words.push(obj);
      if (item > 0) positive.push(obj);
      if (item < 0) negative.push(obj);

      score += item;
    }
  }

  // devolvemos resultados
  var result = {
    score: score,
    comparative: score / tokens.length,
    tokens: tokens,
    words: words,
    positive: positive,
    negative: negative,
    negation: negation
  };

  if (callback === null) return result;
  process.nextTick(function () {
    callback(null, result);
  });
};
